x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -0,0 +1,218 @@
+        synchronized (ioLock) {
+            savedIoControl = ioControl;
+        }
+    }
+
+    private void resumeProducer() {
+        synchronized (ioLock) {
+            if (savedIoControl != null) {
+                savedIoControl.requestInput();
+                savedIoControl = null;
+            }
+        }
+    }
+
+    // if there is currently a thread running in a loop, it should pick up this new request
+    // if not, check if this thread is one of ours and reuse it
+    // else, offload to a new thread so we do not block Apache's connection thread or ElasticSearch's transport thread
+    private void runOrScheduleNextMessage() {
+        if (subscriber != null && isRunning.compareAndSet(false, true)) {
+            var currentThreadPool = EsExecutors.executorName(Thread.currentThread().getName());
+            if (UTILITY_THREAD_POOL_NAME.equalsIgnoreCase(currentThreadPool)) {
+                new OffloadThread().run();
+            } else {
+                threadPool.executor(UTILITY_THREAD_POOL_NAME).execute(new OffloadThread());
+            }
+        }
+    }
+
+    @Override
+    public void responseCompleted(HttpContext httpContext) {}
+
+    @Override
+    public void failed(Exception e) {
+        if (this.isDone.compareAndSet(false, true)) {
+            this.ex = e;
+        }
+        if (isRunning.get() == false && subscriber != null && pendingRequest.compareAndSet(true, false)) {
+            subscriber.onError(ex);
+        }
+    }
+
+    @Override
+    public Exception getException() {
+        return ex;
+    }
+
+    @Override
+    public HttpResponse getResult() {
+        return response;
+    }
+
+    @Override
+    public boolean isDone() {
+        return isDone.get();
+    }
+
+    @Override
+    public void close() {
+        // if isRunning == true, the OffloadThread will close out the subscription
+        // if pendingRequest is false, the next call to request() will close out the subscription
+        if (isDone.compareAndSet(false, true)
+            && isRunning.get() == false
+            && subscriber != null
+            && pendingRequest.compareAndSet(true, false)) {
+            subscriber.onComplete();
+        }
+    }
+
+    @Override
+    public boolean cancel() {
+        resultQueue.clear();
+        close();
+        return true;
+    }
+
+    private class HttpSubscription implements Flow.Subscription {
+        @Override
+        public void request(long n) {
+            pendingRequest.set(true);
+            // always run in case we're waking up from a pause and need to start a new thread
+            runOrScheduleNextMessage();
+        }
+
+        @Override
+        public void cancel() {
+            StreamingHttpResultPublisher.this.cancel();
+        }
+    }
+
+    private class OffloadThread implements Runnable {
+        @Override
+        public void run() {
+            while (resultQueue.isEmpty() == false && pendingRequest.compareAndSet(true, false)) {
```

Comment:
Won't `pendingRequest.compareAndSet(true, false)` always be false after the first loop?

Reply:
`onNext` should eventually call `request` which sets it to true (or now it increases demand), so in theory we should be "stuck" in this loop until we are out of demand or bytes.  In practice we may have breaks in between.